File: src/cowtree/captures.py
"""Freeze private source intent once before submitting it to the authority."""

from dataclasses import dataclass
import os

from pydantic import TypeAdapter

from cowtree.errors import CowtreeError, CowtreeErrorCode
from cowtree.leaves import Leaves
from cowtree.metadata import Metadata, MetadataError
from cowtree.metadata_types import EntryKind, MetadataReason, Operation, Proposal, Receipt, Request
from cowtree.seals import Seals
from cowtree.views import Views
from cowtree.workspace import Workspace
from cowtree.workspace_types import Leaf, Publication


@dataclass(frozen=True)
class Captures:
    workspace: Workspace

    def capture(self, identity: int) -> Publication | None:
        with self.workspace.session():
            leaf = Leaves(self.workspace).read(identity=identity)
            if leaf.check_candidate is not None:
                raise CowtreeError(CowtreeErrorCode.INVALID_ARGUMENTS, "check views cannot publish")
            if leaf.pending is not None:
                raise CowtreeError(
                    CowtreeErrorCode.INVALID_ARGUMENTS,
                    "publication already pending; resume or abort it",
                )
            if Views(self.workspace).current(leaf=leaf) == leaf.origins:
                return None
        node = Seals(self.workspace).seal(identity=identity, retain=False)
        leaf = Leaves(self.workspace).read(identity=identity)
        paths = tuple(
            sorted(
                path
                for path in set(node.source) | set(leaf.origins)
                if node.source.get(path) != leaf.origins.get(path)
            )
        )
        Views(self.workspace).acquire(identity=identity, paths=paths)
        with self.workspace.session() as metadata:
            leaf = Leaves(self.workspace).read(identity=identity)
            if leaf.node != node.id or leaf.pending is not None:
                raise CowtreeError(
                    CowtreeErrorCode.INVALID_ARGUMENTS, "leaf changed during capture"
                )
            changes = {
                path: node.source.get(path)
                for path in paths
                if node.source.get(path) != leaf.origins.get(path)
            }
            if not changes:
                return None
            pending = Publication(
                request=Request(leaf=identity, sequence=leaf.sequence),
                node=node.id,
                changes=changes,
            )
            leaf = leaf.model_copy(update={"pending": pending})
            Leaves(self.workspace).save(leaf=leaf)
            result = self.submit(metadata=metadata, leaf=leaf)
        return result

    def submit(self, metadata: Metadata, leaf: Leaf) -> Publication:
        pending = leaf.pending
        assert pending is not None
        node = self.workspace.nodes.read(identity=pending.node)
        self.workspace.nodes.verify(node=node)
        for path, entry in pending.changes.items():
            grant = leaf.grants[path]
            if entry is not None:
                source = self.workspace.root / "nodes" / node.id / "tree" / path
                if entry.kind is EntryKind.SYMLINK:
                    staged = metadata.call(
                        Operation.STAGE,
                        {"leaf": leaf.id, "data": list(os.fsencode(os.readlink(source)))},
                    )
                else:
                    staged = metadata.call(
                        Operation.STAGE_FILE, {"leaf": leaf.id, "path": str(source)}
                    )
                if staged.decode("object", TypeAdapter(str)) != entry.object:
                    raise CowtreeError(CowtreeErrorCode.COMMAND_FAILED, "captured object changed")
            metadata.call(
                Operation.EDIT,
                {
                    "grant": grant.model_dump(mode="json"),
                    "value": None if entry is None else entry.model_dump(mode="json"),
                },
            )
        proposal = metadata.call(
            Operation.PROPOSE,
            {
                "input": {
                    "request": pending.request.model_dump(mode="json"),
                    "paths": sorted(pending.changes),
                }
            },
        ).decode("proposal", TypeAdapter(Proposal))
        if {change.path: change.value for change in proposal.changes} != pending.changes:
            raise CowtreeError(
                CowtreeErrorCode.COMMAND_FAILED, "request identity disagrees with captured source"
            )
        submitted = pending.model_copy(update={"submitted": True})
        Leaves(self.workspace).save(leaf=leaf.model_copy(update={"pending": submitted}))
        return submitted

    def abort(self, identity: int) -> None:
        with self.workspace.session(recover=False) as metadata:
            Views(self.workspace).recover_locked(metadata=metadata)
            Seals(self.workspace).recover_locked(metadata=metadata)
            Leaves(self.workspace).recover_locked(metadata=metadata)
            leaf = Leaves(self.workspace).read(identity=identity)
            pending = leaf.pending
            if pending is None:
                raise CowtreeError(CowtreeErrorCode.INVALID_ARGUMENTS, "no publication is pending")
            if not pending.aborting:
                try:
                    receipt = metadata.call(
                        Operation.RESULT, {"request": pending.request.model_dump(mode="json")}
                    ).decode("result", TypeAdapter(Receipt | None))
                except MetadataError as error:
                    if error.reason is not MetadataReason.REQUEST_EXPIRED or pending.submitted:
                        raise
                    receipt = None
                if receipt is not None:
                    raise CowtreeError(
                        CowtreeErrorCode.INVALID_ARGUMENTS,
                        "publication already committed; recover its receipt",
                    )
                pending = pending.model_copy(update={"aborting": True})
                leaf = leaf.model_copy(update={"pending": pending})
                Leaves(self.workspace).save(leaf=leaf)
            self.cancel(metadata=metadata, leaf=leaf)

    def cancel(self, metadata: Metadata, leaf: Leaf) -> None:
        pending = leaf.pending
        assert pending is not None
        assert pending.aborting
        try:
            metadata.call(Operation.ABORT, {"request": pending.request.model_dump(mode="json")})
        except MetadataError as error:
            if error.reason is not MetadataReason.REQUEST_EXPIRED:
                raise
        for entry in pending.changes.values():
            if entry is not None:
                metadata.call(Operation.DISCARD_UPLOAD, {"leaf": leaf.id, "object": entry.object})
        Leaves(self.workspace).save(
            leaf=leaf.model_copy(update={"pending": None, "sequence": pending.request.sequence + 1})
        )

    def recover_locked(self, metadata: Metadata) -> None:
        for leaf in Leaves(self.workspace).records():
            if leaf.pending is not None and leaf.pending.aborting:
                self.cancel(metadata=metadata, leaf=leaf)
            elif leaf.pending is not None and not leaf.pending.submitted:
                self.submit(metadata=metadata, leaf=leaf)

File: crates/workspace/src/captures.rs
// Copyright (c) 2026 Windsor Nguyen

//! Freeze publication intent once and replay submission under the same request identity.

use std::{collections::BTreeMap, fs, os::unix::ffi::OsStrExt};

use cowtree_metadata::{FileKind, LeafId, ProposalInput, RequestId};

use crate::{Error, Leaf, Pending, Result, Workspace, error::Issue, session::Session};

impl Workspace {
    pub fn capture(&self, identity: LeafId) -> Result<Option<Pending>> {
        self.lock()?.capture(identity)
    }

    pub fn abort(&self, identity: LeafId) -> Result<()> {
        let mut session = self.lock_raw()?;
        session.recover_views()?;
        session.recover_seals()?;
        session.recover_forks()?;
        session.pin_origins()?;
        let mut leaf = session.read_leaf(identity)?;
        let pending =
            leaf.pending.as_mut().ok_or_else(|| Issue::PendingPublication.at(&leaf.path))?;
        if !pending.aborting {
            let receipt = match session.authority.result(pending.request) {
                Ok(receipt) => receipt,
                Err(cowtree_metadata::Error::RequestExpired(_)) if !pending.submitted => None,
                Err(error) => return Err(error.into()),
            };
            if receipt.is_some() {
                return Err(Issue::InvalidCandidate.at(&leaf.path));
            }
            pending.aborting = true;
            session.save_leaf(&leaf)?;
        }
        session.cancel_capture(leaf)?;
        session.pin_origins()
    }
}

impl Session {
    pub(crate) fn capture(&mut self, identity: LeafId) -> Result<Option<Pending>> {
        let leaf = self.read_leaf(identity)?;
        if leaf.check_candidate.is_some() {
            return Err(Issue::InvalidCandidate.at(&leaf.path));
        }
        if leaf.pending.is_some() {
            return Err(Issue::PendingPublication.at(&leaf.path));
        }
        if self.current(&leaf)? == leaf.origins {
            return Ok(None);
        }
        let node = self.seal(identity, crate::Retention::Referenced)?;
        let selected: std::collections::BTreeSet<_> = node
            .source
            .keys()
            .chain(leaf.origins.keys())
            .filter(|path| node.source.get(*path) != leaf.origins.get(*path))
            .cloned()
            .collect();
        let selected: Vec<_> = selected.into_iter().collect();
        let mut leaf = self.acquire(identity, &selected)?;
        if leaf.node != node.id || leaf.pending.is_some() {
            return Err(Issue::PendingPublication.at(&leaf.path));
        }
        let changes = selected
            .into_iter()
            .filter(|path| node.source.get(path) != leaf.origins.get(path))
            .map(|path| {
                let value = node.source.get(&path).cloned();
                (path, value)
            })
            .collect::<BTreeMap<_, _>>();
        if changes.is_empty() {
            return Ok(None);
        }
        leaf.pending = Some(Pending {
            request: RequestId { leaf: identity, sequence: leaf.sequence.get() },
            node: node.id,
            changes,
            submitted: false,
            aborting: false,
            candidate: None,
            batch: None,
            parent_node: None,
            validation: None,
        });
        self.save_leaf(&leaf)?;
        crate::fault::checkpoint("after-capture-intent");
        let pending = self.submit_capture(leaf)?;
        self.pin_origins()?;
        Ok(Some(pending))
    }

    pub fn submit_capture(&mut self, mut leaf: Leaf) -> Result<Pending> {
        let pending =
            leaf.pending.as_ref().ok_or_else(|| Issue::IncompleteRecord.at(&leaf.path))?;
        let node = self.nodes().read(&pending.node)?;
        self.nodes().verify(&node)?;
        for (path, entry) in &pending.changes {
            let grant =
                leaf.grants.get(path).ok_or_else(|| Issue::IncompleteRecord.at(&leaf.path))?;
            if let Some(entry) = entry {
                let source = self
                    .workspace
                    .root
                    .join("nodes")
                    .join(node.id.as_str())
                    .join("tree")
                    .join(path.as_str());
                let staged = if entry.kind.file_kind() == FileKind::Symlink {
                    let link = fs::read_link(&source).map_err(|error| Error::io(&source, error))?;
                    self.authority.stage(leaf.id, link.as_os_str().as_bytes())?
                } else {
                    self.authority.stage_file(leaf.id, &source)?
                };
                if staged != entry.object {
                    return Err(Issue::SourceChanged.at(&source));
                }
            }
            self.authority.edit(grant, entry.clone())?;
        }
        let proposal = self.authority.propose(ProposalInput {
            request: pending.request,
            paths: pending.changes.keys().cloned().collect(),
        })?;
        let changes: BTreeMap<_, _> =
            proposal.changes.into_iter().map(|change| (change.path, change.value)).collect();
        if changes != pending.changes {
            return Err(Issue::IncompleteRecord.at(&leaf.path));
        }
        let mut submitted = pending.clone();
        submitted.submitted = true;
        leaf.pending = Some(submitted.clone());
        self.save_leaf(&leaf)?;
        Ok(submitted)
    }

    pub fn cancel_capture(&mut self, mut leaf: Leaf) -> Result<()> {
        let pending = leaf
            .pending
            .as_ref()
            .filter(|pending| pending.aborting)
            .ok_or_else(|| Issue::IncompleteRecord.at(&leaf.path))?;
        match self.authority.abort(pending.request) {
            Ok(()) | Err(cowtree_metadata::Error::RequestExpired(_)) => {}
            Err(error) => return Err(error.into()),
        }
        for entry in pending.changes.values().flatten() {
            self.authority.discard_upload(leaf.id, &entry.object)?;
        }
        leaf.sequence = std::num::NonZeroU64::new(
            pending
                .request
                .sequence
                .checked_add(1)
                .ok_or(cowtree_metadata::Error::CounterExhausted)?,
        )
        .ok_or(cowtree_metadata::Error::CounterExhausted)?;
        leaf.pending = None;
        self.save_leaf(&leaf)
    }

    pub fn recover_captures(&mut self) -> Result<()> {
        for leaf in self.leaves()? {
            match &leaf.pending {
                Some(pending) if pending.aborting => self.cancel_capture(leaf)?,
                Some(pending) if !pending.submitted => {
                    self.submit_capture(leaf)?;
                }
                _ => {}
            }
        }
        Ok(())
    }
}
